mesonbuild/modules/qt.py

```diff
@@ -428,22 +428,35 @@ def preprocess(self, state: 'ModuleState', args: T.List[T.Union[str, File]], kwa
                 if not isinstance(args[0], str):
                     raise build.InvalidArguments('First argument to qt.preprocess must be a string')
                 rcc_kwargs['name'] = args[0]
-            sources.extend(self.compile_resources(state, tuple(), rcc_kwargs).return_value)
+            sources.append(self.compile_resources(state, tuple(), rcc_kwargs).return_value)
 
         if kwargs['ui_files']:
             ui_kwargs: 'UICompilerKwArgs' = {'sources': kwargs['ui_files'], 'extra_args': kwargs['uic_extra_arguments'], 'method': method}
-            sources.extend(self.compile_ui(state, tuple(), ui_kwargs).return_value)
+            sources.append(self.compile_ui(state, tuple(), ui_kwargs).return_value)
 
         if kwargs['moc_headers'] or kwargs['moc_sources']:
+            # Preprocess allows string vlues for include_directories, but
+            # compile_moc doesn't. The preprocess interface is quite old, and
+            # modern interfaces generally don't allow strings like this.
+            string_inc_dirs: T.List[str] = []
+            include_dirs: T.List[IncludeDirsHolder] = []
+            for i in kwargs['include_directories']:
+                if isinstance(i, str):
+                    string_inc_dirs.append(i)
+                else:
+                    include_dirs.append(i)
+            if string_inc_dirs:
```

Comment:
Not sure how relevant this is, but this changes the ordering of the include directories, which might be significant.